parse/grammar/KuneiformParser.g4

```diff
@@ -0,0 +1,378 @@
+    | (PLUS | MINUS)? DIGITS_ PERIOD DIGITS_ # decimal_literal
+    | (TRUE | FALSE) # boolean_literal
+    | NULL # null_literal
+    | BINARY_ # binary_literal
+;
+
+identifier_list:
+    IDENTIFIER (COMMA IDENTIFIER)*
+;
+
+type:
+    IDENTIFIER (LPAREN DIGITS_ COMMA DIGITS_ RPAREN)? (LBRACKET RBRACKET)? // Handles arrays of any type, including nested arrays
+;
+
+type_cast:
+    TYPE_CAST type
+;
+
+variable:
+    VARIABLE | CONTEXTUAL_VARIABLE
+;
+
+variable_list:
+    variable (COMMA variable)*
+;
+
+/*
+    The following section includes parser rules for top-level Kuneiform.
+    These are the rules that parse the schema / DDL, and are used pre-consensus.
+*/
+
+// schema is the parser entrypoint for an entire
+// Kuneiform schema.
+schema:
+    database_declaration
+    (use_declaration | table_declaration
+     | action_declaration | procedure_declaration
+     | foreign_procedure_declaration
+    )*
+;
+
+annotation:
+    // sort've a hack; annotations don't technically use contextual variables, but they have
+    // the same syntax of @identifier
+    CONTEXTUAL_VARIABLE LPAREN (IDENTIFIER EQUALS literal (COMMA IDENTIFIER EQUALS literal)*)? RPAREN
+;
+
+database_declaration:
+    DATABASE IDENTIFIER SCOL
+;
+
+use_declaration:
+    USE IDENTIFIER
+    (LBRACE IDENTIFIER COL literal (COMMA IDENTIFIER COL literal)* RBRACE)?
+    AS IDENTIFIER SCOL
+;
+
+table_declaration:
+     TABLE IDENTIFIER LBRACE
+     column_def (COMMA (column_def | index_def | foreign_key_def))*
+     RBRACE
+ ;
+
+column_def:
+    name=IDENTIFIER type constraint*
+;
+
+index_def:
+    HASH_IDENTIFIER
+    (UNIQUE | INDEX | PRIMARY)
+    LPAREN  columns=identifier_list RPAREN
+;
+
+foreign_key_def:
+    (FOREIGN KEY|LEGACY_FOREIGN_KEY) // for backwards compatibility
+    LPAREN child_keys=identifier_list RPAREN
+    (REFERENCES|REF) parent_table=IDENTIFIER LPAREN parent_keys=identifier_list RPAREN
+    foreign_key_action*
+;
+
+// variability here is to support legacy syntax
+foreign_key_action:
+    ((ON UPDATE|LEGACY_ON_UPDATE)|(ON DELETE|LEGACY_ON_DELETE)) DO? ((NO ACTION|LEGACY_NO_ACTION)|CASCADE|(SET NULL|LEGACY_SET_NULL)|(SET DEFAULT|LEGACY_SET_DEFAULT)|RESTRICT)
+;
+
+type_list:
+    type (COMMA type)*
+;
+
+named_type_list:
+    IDENTIFIER type (COMMA IDENTIFIER type)*
+;
+
+typed_variable_list:
+    variable type (COMMA variable type)*
+;
+
+constraint:
+    MIN LPAREN literal RPAREN # MIN
+    | MAX LPAREN literal RPAREN # MAX
+    | MINLEN LPAREN literal RPAREN # MIN_LEN
+    | MAXLEN LPAREN literal RPAREN # MAX_LEN
+    | (NOTNULL|NOT NULL) # NOT_NULL
+    | (LEGACY_PRIMARY_KEY|PRIMARY KEY?) # PRIMARY_KEY
+    | DEFAULT LPAREN literal RPAREN # DEFAULT
+    | UNIQUE # UNIQUE
+;
+
+access_modifier:
+    PUBLIC | PRIVATE | VIEW | OWNER
+;
+
+action_declaration:
+    annotation*
+    ACTION IDENTIFIER
+    LPAREN variable_list? RPAREN
+    (access_modifier)+
+    LBRACE action_block RBRACE
+;
+
+procedure_declaration:
```

Comment:
Yeah Truflation ran into this. Not really something we can handle in the parser. But yeah, it makes me think we need some better access control.

Reply:
not sure how we can ensure this, since we cannot enforce the caller procedure has same annotation as called procedure(or only enforce authn annotation), it's not kwild's concern